### PR TITLE
New maps

### PR DESCRIPTION
**Pull request by bmcfee**
This PR implements #93 -- iterator caching.

It also factors out the randomstate logic to a `util` function, since it now gets used in a few places.

### CODE REVIEW

**Comment by bmcfee** [comment]
@ejhumphrey Can you chime in on this? I think we should not implement the validation function, so this is really just about the cache map.  I mostly want to know if this matches what you had in mind for this feature.

File: pescador/maps.py [hunk 1]
```diff
@@ -236,3 +238,84 @@ def keras_tuples(stream, inputs=None, outputs=None):
             yield (x, y)
         except TypeError:
             raise DataError("Malformed data stream: {}".format(data))
+
+
+def cache(stream, n_cache, prob=0.5, random_state=None):
+    '''Stochastic stream caching.
+
+    With probability `prob`: yield a new item from `stream` and place it in the cache
```
Comment (ejhumphrey) [inline pescador/maps.py]: this means the probability of the predicate happening, right? lower probs tend toward newer stuff, higher probs tend toward older stuff?

Comment (bmcfee) [reply pescador/maps.py]: It's the probability of getting a new item rather than one you've seen before (as described below).

Comment (ejhumphrey) [reply pescador/maps.py]: reading the rest of the docstring clarifies this ... perhaps it's enough for this docstring to speak to the intent of what's meant by caching here? something like "this is useful for effectively speeding up streaming by reusing in-memory datapoints, and should have a smoothing effect on gradients in online learning settings." the latter part is intuitive/hypothetical, but you know what I mean. maybe.

File: pescador/maps.py [hunk 2]
```diff
@@ -236,3 +238,84 @@ def keras_tuples(stream, inputs=None, outputs=None):
...
+    ------
+    data
+        elements of `stream`
+    '''
+
+    if n_cache <= 0:
+        raise PescadorError('n_cache={} must be a positive integer'.format(n_cache))
+
+    if not 0 < prob <= 1:
+        raise PescadorError('prob={} must be a number between 0 and 1.'.format(prob))
```
Comment (ejhumphrey) [inline pescador/maps.py]: maybe "in range (0, 1]"? it can be 1 but not zero, right?

Comment (bmcfee) [reply pescador/maps.py]: :+1:

File: pescador/maps.py [hunk 3]
```diff
@@ -236,3 +238,84 @@ def keras_tuples(stream, inputs=None, outputs=None):
...
+                item = next(stream)
+                yield item
+            except StopIteration:
+                break
+            data_cache.append(item)
+
+        else:
+            # Otherwise, the cache is now full.
+            # Pick a random index
+            idx = rng.randint(low=0, high=n_cache)
```
Comment (ejhumphrey) [inline pescador/maps.py]: didn't we find once upon a time that generating random values one at a time is slow in large quantities? how often does this step get called, dya think? would it be premature to have a large array and shuffle it?

File: pescador/util.py [hunk 4]
```diff
@@ -148,3 +150,34 @@ def batch_length(batch):
             raise PescadorError('Unequal field lengths')
 
     return n
+
+
+def get_rng(random_state):
```
Comment (ejhumphrey) [inline pescador/util.py]: 👌

**Comment by ejhumphrey** [comment]
lgtm, coupla tiny things

File: pescador/maps.py (hunk 3 above)
Comment (bmcfee) [reply pescador/maps.py]: We did, but I think that problem is above the pay grade of this PR.  If we solve that, I'd prefer to do it at one level up of abstraction and across the entire library.

One potential solution to that would be to make a generator that pre-allocates a block of random numbers, yields them one at a time, and then replenishes when necessary.  If that can be done in numba, it should have pretty minimal overhead.

**Comment by bmcfee** [comment]
Thanks for the CR!  Latest commit integrates your suggestions.  I want to punt on RNG efficiency until later.  I'll merge once this passes CI.

**Pull request merged**